README.md
# Vici::ExchangeRates

> A Simple and Rich for Exchange Rates

![CI](https://github.com/maful/vici-exchange-rates/workflows/CI/badge.svg?branch=master)
[![Gem Version](https://badge.fury.io/rb/vici-exchange-rates.svg)](https://badge.fury.io/rb/vici-exchange-rates)
[![GitHub license](https://img.shields.io/github/license/maful/vici-exchange-rates)](https://github.com/maful/vici-exchange-rates/blob/master/LICENSE.txt)

**Vici::ExchangeRates** is an unofficial SDK for the [ExchangeRatesAPI](https://exchangeratesapi.io). Exchange rates API is a free service for current and historical foreign exchange rates [published by the European Central Bank](https://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html).

### Note
- This app requires JSON. If you're using JRuby < 1.7.0
  you'll need to add `gem "json"` to your Gemfile or similar.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'vici-exchange-rates'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install vici-exchange-rates

## Usage

Currency code must be one of the [supported currency codes](https://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html).

**Basic usage**

```ruby
require 'vici/exchange_rates'

rates = Vici::ExchangeRates.new
# or you can retrieve JSON response
rates = Vici::ExchangeRates.new(true) 
rates.fetch
```

**Historical rates for any day since 1999**

```ruby
rates.at_date('2018-02-02')
rates.fetch
# OR
rates.at_date('2018-02-02').fetch 
```

**Latest exchange rates with different base currency (Euro by default)**

```ruby
rates.base_currency('IDR')
rates.fetch 
```

**Specific exchange rates**

```ruby
rates.rates('IDR')

# multi exchange rates
rates.rates(['IDR', 'USD']) 
rates.fetch 
# => {"rates"=>{"USD"=>1.1052, "IDR"=>15091.51}, "base"=>"EUR", "date"=>"2020-01-31"}
```

**Historical rates for a time period**

```ruby
rates.period(date_from: '2018-01-03', date_to: '2018-02-05') 
rates.fetch 
```

**Limit to specific exchange rates**

```ruby
rates.rates(['IDR', 'JPY']).period(date_from: '2018-01-03', date_to: '2018-02-05') 
rates.fetch 
```

**Historical rates with different currency**

```ruby
rates.base_currency('JPY').period(date_from: '2018-01-03', date_to: '2018-02-05') 
rates.fetch
```

## Supported Currencies

The library supports any currency currently available on the European Central Bank's web service, please refer to [here](https://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html).

## Contributing

1. Fork [the repo](https://github.com/maful/vici-exchange-rates)
2. Grab dependencies: `bundle install`
3. Make sure everything is working: `bundle exec rake spec`
4. Make your changes
5. Test your changes
5. Create a Pull Request
6. Celebrate!!!!!

This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## Copyright

Copyright (c) 2020 Maful Prayoga A. See LICENSE for further details.


CHANGELOG.md
# Changelog

## 0.1.1
- Add Github Actions
- Fix typo
- Add badges

## 0.1.0
- First gem release

spec/spec_helper.rb
require "bundler/setup"
require "vici/exchange_rates"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/vici/exchange_rates_spec.rb
RSpec.describe Vici::ExchangeRates do
  subject(:lookup) { Vici::ExchangeRates.new }

  it 'has a version number' do
    expect(Vici::ExchangeRates::VERSION).not_to be nil
  end

  it 'has a BASE URI' do
    expect(Vici::ExchangeRates::BASE_URI).to eq('https://api.exchangeratesapi.io')
  end

  describe 'valid argument' do
    context 'get latest rates' do
      subject { lookup.fetch }

      it 'should base currency is EUR' do
        expect(subject['base']).to eq('EUR')
      end

      it 'should rates is not empty' do
        expect(subject['rates']).not_to be_empty
      end

      it 'should response hash' do
        expect(subject).to be_kind_of(Hash)
      end
    end

    context 'get rates at the day' do
      let(:date) { '2010-01-22' }
      subject { lookup.at_date(date).fetch }

      it 'should base currency is EUR' do
        expect(subject['base']).to eq('EUR')
      end

      it 'should rates is not empty' do
        expect(subject['rates']).not_to be_empty
      end

      it 'should date is valid' do
        expect(subject['date']).to eq(date)
      end

      it 'should response hash' do
        expect(subject).to be_kind_of(Hash)
      end
    end

    context 'get latest rates with different currency' do
      subject { lookup.base_currency('IDR').fetch }

      it 'should base currency is IDR' do
        expect(subject['base']).to eq('IDR')
      end

      it 'should rates is not empty' do
        expect(subject['rates']).not_to be_empty
      end

      it 'should response hash' do
        expect(subject).to be_kind_of(Hash)
      end
    end

    context 'get latest with specific exchange rates' do
      let(:currencies) { %w[IDR USD] }
      subject { lookup.rates(currencies).fetch }

      it 'should base currency is EUR' do
        expect(subject['base']).to eq('EUR')
      end

      it 'should length rates is valid' do
        expect(subject['rates'].length).to eq(currencies.length)
      end

      it 'should rates has key' do
        expect(subject['rates'].has_key?(currencies.first)).to be true
        expect(subject['rates'].has_key?(currencies[1])).to be true
      end

      it 'should response hash' do
        expect(subject).to be_kind_of(Hash)
      end
    end

    context 'get historical rates for a time period' do
      let(:date_from) { '2019-01-01' }
      let(:date_to) { '2019-01-03' }
      subject { lookup.period(date_from: date_from, date_to: date_to).fetch }

      it 'should base currency is EUR' do
        expect(subject['base']).to eq('EUR')
      end

      it 'should rates is not empty' do
        expect(subject['rates']).not_to be_empty
      end

      it 'should value start_at is valid' do
        expect(subject['start_at']).to eq(date_from)
      end

      it 'should value end_at is valid' do
        expect(subject['end_at']).to eq(date_to)
      end

      it 'should response hash' do
        expect(subject).to be_kind_of(Hash)
      end

      context 'with specific exchange rates' do
        subject do
          lookup.rates(%w[JPY USD])
              .period(date_from: date_from, date_to: date_to)
              .fetch
        end

        it 'should length rates is valid' do
          expect(subject['rates'].length).to eq(2)
        end
      end

      context 'with different base currency' do
        subject do
          lookup.base_currency('JPY')
              .period(date_from: date_from, date_to: date_to)
              .fetch
        end

        it 'should base currency is JPY' do
          expect(subject['base']).to eq('JPY')
        end
      end
    end

    context 'response json format' do
      subject { Vici::ExchangeRates.new(true).fetch }

      it 'should response json' do
        expect(subject).to be_kind_of(String)
      end
    end
  end

  describe 'invalid argument' do
    context 'when invalid date format' do
      subject { lookup.at_date('2010').fetch }

      it 'raises' do
        expect { subject }.to raise_error(ArgumentError, 'The specified date is invalid. Please use ISO 8601 notation (e.g. YYYY-MM-DD).')
      end
    end

    context 'when invalid currency notation' do
      subject { lookup.base_currency('rupiah').fetch }

      it 'raises' do
        expect { subject }.to raise_error('The specified currency code is invalid. Please use ISO 4217 notation (e.g. IDR).')
      end
    end

    context 'when currency not supported' do
      subject { lookup.base_currency('ida').fetch }

      it 'raises' do
        expect { subject }.to raise_error('The specified currency code is not currently supported.')
      end
    end
  end
end


lib/vici/exchange_rates.rb
require 'active_support/core_ext/object'
require 'faraday'
require 'json'

require 'vici/exchange_rates/version'

module Vici
  class ExchangeRates
    class Error < StandardError; end

    BASE_URI = 'https://api.exchangeratesapi.io'.freeze
    CURRENCIES = %w[USD GBP EUR JPY BGN CZK DKK HUF PLN RON SEK CHF ISK NOK HRK RUB TRY AUD BRL CAD CNY HKD IDR ILS INR KRW MXN MYR NZD BHP SGD THB ZAR].freeze
    CURRENCY_REGEX = /^[A-Z]{3}$/.freeze
    DATE_REGEX = /^[0-9]{4}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])$/.freeze

    def initialize(json = false)
      @client = Faraday.new BASE_URI
      @options = {}
      @endpoint = 'latest'
      @rates = []
      @date = nil
      @date_from = nil
      @date_to = nil
      @base = nil
      @return_json = json
    end

    def base_currency(code)
      currency_code = sanitize_curreny_code(code)
      verify_currency_code(currency_code)
      @base = code
      self
    end

    def at_date(date)
      @date = validate_date(date)
      self
    end

    def rates(currencies)
      if currencies.is_a?(Array)
        currencies.each do |currency|
          currency_code = sanitize_curreny_code(currency)

          verify_currency_code(currency_code)

          @rates << currency_code
        end
      else
        currency_code = sanitize_curreny_code(currencies)

        verify_currency_code(currency_code)

        @rates << currency_code
      end
      self
    end

    def period(date_from:, date_to:)
      @date_from = validate_date(date_from)
      @date_to = validate_date(date_to)
      self
    end

    def fetch
      if @date_from.present? && @date_to.present?
        @endpoint = 'history'
        @options.merge!(start_at: @date_from, end_at: @date_to)
      elsif @date.present?
        @endpoint = @date
      else
        @endpoint
      end

      @options['base'] = @base if @base.present?

      @options['symbols'] = @rates.join(',') if @rates.length.positive?

      request = @client.get(@endpoint, @options)
      response = JSON.parse(request.body)

      if @return_json
        JSON.generate(response)
      else
        response
      end
    end

    private

    def verify_currency_code(code)
      currency_code = sanitize_curreny_code(code)

      unless currency_valid?(currency_code)
        raise 'The specified currency code is invalid. Please use ISO 4217 notation (e.g. IDR).'
      end

      unless currency_supported?(currency_code)
        raise 'The specified currency code is not currently supported.'
      end
    end

    def currency_valid?(code)
      if code.present?
        return false if code.length != 3

        return false unless CURRENCY_REGEX.match?(code)

        return true
      end
      false
    end

    def currency_supported?(code)
      CURRENCIES.include?(code)
    end

    def validate_date(date)
      unless date_valid?(date)
        raise ArgumentError, 'The specified date is invalid. Please use ISO 8601 notation (e.g. YYYY-MM-DD).'
      end

      date
    end

    def date_valid?(date)
      return DATE_REGEX.match?(date) if date.present?

      false
    end

    def sanitize_curreny_code(code)
      code.strip.upcase
    end
  end
end
